libs/tour/src/lib/services/tour-service/tour.service.ts

```diff
@@ -0,0 +1,486 @@
+	private currentDirection: NgxTourDirection;
+
+	/**
+	 * The currently active step in the tour as a subject
+	 */
+	private readonly currentStepSubject: BehaviorSubject<NgxTourStep> =
+		new BehaviorSubject<NgxTourStep>(undefined);
+
+	/**
+	 * The previously active step in the tour as a subject
+	 */
+	private readonly previousStepSubject: BehaviorSubject<NgxTourStep> =
+		new BehaviorSubject<NgxTourStep>(undefined);
+
+	/**
+	 * The index of the current step of the tour as a subject
+	 */
+	private readonly currentIndexSubject: BehaviorSubject<number> = new BehaviorSubject<number>(0);
+
+	/**
+	 * Whether the tour has started as a subject
+	 */
+	private readonly tourStartedSubject: Subject<void> = new Subject();
+
+	/**
+	 * Whether the tour has ended as a subject
+	 */
+	private readonly tourEndedSubject: Subject<void> = new Subject();
+
+	/**
+	 * The currently active tour as a subject
+	 */
+	private readonly currentTourSubject: BehaviorSubject<NgxTourStep[]> = new BehaviorSubject<
+		NgxTourStep[]
+	>(undefined);
+
+	/**
+	 * The start scroll position of the page
+	 */
+	private startingScrollPosition: number;
+
+	/**
+	 * The currently active tour
+	 */
+	public readonly currentTour$: Observable<NgxTourStep[]> =
+		this.currentTourSubject.asObservable();
+
+	/**
+	 * The currently active step
+	 */
+	public readonly currentStep$: Observable<NgxTourStep> = this.currentStepSubject
+		.asObservable()
+		.pipe(distinctUntilChanged(), filter(Boolean));
+
+	/**
+	 * The currently active index
+	 */
+	public readonly currentIndex$: Observable<number> = this.currentIndexSubject
+		.asObservable()
+		.pipe(distinctUntilChanged());
+
+	/**
+	 * The previously active step
+	 */
+	public readonly previousStep$: Observable<NgxTourStep> = this.previousStepSubject
+		.asObservable()
+		.pipe(distinctUntilChanged(), filter(Boolean));
+
+	/**
+	 * Whether the tour has ended
+	 */
+	public readonly tourEnded$: Observable<void> = this.tourEndedSubject.asObservable();
+
+	/**
+	 * Whether the tour has started
+	 */
+	public readonly tourStarted$: Observable<void> = this.tourStartedSubject.asObservable();
+
+	constructor(
+		private readonly cdkOverlayService: Overlay,
+		@Inject(PLATFORM_ID) private readonly platformId: string,
+		@Inject(NgxTourStepToken) private readonly component: Type<NgxTourStepComponent>
+	) {}
+
+	/**
+	 * Start a provided tour
+	 *
+	 * @param tour - The tour we wish to start
+	 * @param onClose - An optional on close function we wish to run
+	 * @param startIndex - An optional index we wish to start the tour from , default this is 0
+	 */
+	public startTour(
+		tour: NgxTourStep[],
+		onClose?: NgxTourAction,
+		startIndex = 0
+	): Observable<NgxTourInteraction> {
+		// Iben: Save the current scroll position so we can return to it when we close the tour
+		this.runInBrowser(() => {
+			this.startingScrollPosition = window.scrollY;
+		});
+
+		// Iben: Loop over the tour and set subjects for all elements that haven't been found yet
```

Comment:
Maybe switch to `reduce` to create a new array each time and get rid of the reference.
```typescript
  // Iben: Loop over the tour and set subjects for all elements that haven't been found yet
  this.elements = tour.reduce((acc, step) => {
    if (!this.elements[step.tourItem]) {
      return {
        ...acc,
       [step.tourItem] = new BehaviorSubject<NgxTourItemDirective>(undefined);
      };

      return acc;
    }
  });
```

Reply:
After internal discussions we decided to keep this as is.